common/base/include/claragenomics/types.hpp

```diff
@@ -12,6 +12,12 @@
 
 #include <cstdint>
 
+#ifdef __cpp_lib_optional
```

Comment:
Interestingly enough even GCC 5.4 seems to support it.
As you said, `__has_include` is C++17 and as we are testing for a feature added in C++17 that also isn't totally fool proof, right?

Reply:
This seems more complicated than I thought.
Originally, I thought `__has_include` is preferable to `__cpp_lib_optional`, because for the latter C++17 compilers may just silently fall back to experimental/optional, even though they do support optional natively, since they don't have the `__cpp_lib_optional` macro yet.

Then I thought your solution is actually preferable, because `#if __has_include(<optional>)` will just not compile with pre-C++-17 compilers and therefore defy it's purpose here.

However, then I noticed that the `__cpp_lib_*` are actually defined in the C++20 `<version>` header. This means the code as it is will most likely silently fall back to the `experimental/optional` even for C++20 compilers.

Obviously, just including `<version>` and be done with it would be too easy... The `<version>` header does not exist for "older" compilers (eg. GCC <=8.3.0), so we cannot just include that, but need even more macro-fu to achieve what we want here.

How about a simple `#if __cplusplus >= 201703` instead?

Reply:
Yes, let's simply go for `#if __cplusplus >= 201703`

Reply:
sounds good.